<small>

## 🔍 RAG LLM for E-Commerce Product Analysis

### 🚀 Objective
This notebook demonstrates how to build a **Retrieval-Augmented Generation (RAG)** system — a cutting-edge approach that combines **real data** with the power of **LLMs (Large Language Models)**.

Using real-world US e-commerce transaction data from 2020, we can answer business-relevant questions like:

> _"Which products offer high discounts but still generate profit?"_  
> _"Which categories tend to be most unprofitable?"_

---

### 🧠 How Does It Work?

🔹 **1. Load and process product data**  
We select key information: product name, category, sales, discount, and profit.

🔹 **2. Turn text into embeddings**  
Using `SentenceTransformers`, each product row is transformed into a semantic vector (embedding) that captures its meaning.

🔹 **3. Build a fast search index using FAISS**  
When a user asks a question, the system searches for the most relevant product data using similarity matching.

🔹 **4. LLM (GPT) generates a contextual answer**  
The LLM reads the retrieved data and answers intelligently based on that context.

---

### 🎯 What Can You Ask?

You can explore the data with natural language queries like:

- **"Which product had the highest discount?"**
- **"Which categories bring the highest profit?"**
- **"Are there discounted products that still made money?"**

> ❗ Avoid questions the dataset can’t answer (e.g., *"Which items were paid in cash?"*) — payment data is not included.

---

### 💼 Why It’s Powerful

- 🔎 **Instant insights from large datasets**
- 🧠 **Combines retrieval and reasoning in one pipeline**
- 📊 **Can be extended into chatbots, dashboards, or data assistants**

---

### 🧩 Dataset Focus
The dataset contains 18+ columns, but our RAG pipeline focuses on:
- `Product Name`
- `Category` & `Sub-Category`
- `Sales` | `Discount` | `Profit`

---

### 📌 Tech Stack
- 🐼 `pandas` & `numpy`
- 🧠 `sentence-transformers` (MiniLM)
- 🔎 `faiss` (Facebook AI Similarity Search)
- 🤖 `OpenAI GPT-4` (via API)


<small>

<small>

### Step 1: Preparing Tabular Data

To enable the LLM to generate relevant answers based on real information, we first need to prepare the data in a **tabular format**.

Tabular data refers to structured datasets organized in **rows and columns**, where each row represents a record (e.g., a customer), and each column represents a feature or attribute (e.g., income, credit amount, age).

The format of the data can vary — commonly used formats include:
- `.csv` (Comma-Separated Values)
- `.xlsx` (Excel files)
- SQL tables
- Other structured data sources

> Well-structured tabular data is essential for effective retrieval and contextual gen

<small>

In [3]:
import pandas as pd
from sentence_transformers import SentenceTransformer, util
import faiss
import numpy as np
import openai

# 1. BACA FILE BARU
df = pd.read_csv("data/US  E-commerce records 2020.csv", encoding='ISO-8859-1')
df.head()
# 2. CEK NAMA KOLOM
# df.columns = df.columns.str.strip()

Order Date  Row ID        Order ID       Ship Mode Customer ID      Segment  \
0   01-01-20     849  CA-2017-107503  Standard Class    GA-14725     Consumer   
1   01-01-20    4010  CA-2017-144463  Standard Class    SC-20725     Consumer   
2   01-01-20    6683  CA-2017-154466     First Class    DP-13390  Home Office   
3   01-01-20    8070  CA-2017-151750  Standard Class    JM-15250     Consumer   
4   01-01-20    8071  CA-2017-151750  Standard Class    JM-15250     Consumer   

         Country         City       State  Postal Code   Region  \
0  United States       Lorain        Ohio        44052     East   
1  United States  Los Angeles  California        90036     West   
2  United States     Franklin   Wisconsin        53132  Central   
3  United States   Huntsville       Texas        77340  Central   
4  United States   Huntsville       Texas        77340  Central   

        Product ID         Category Sub-Category  \
0  FUR-FU-10003878        Furniture  Furnishings   
1  FUR-FU-10001215        Furniture  Furnishings   
2  OFF-BI-10002012  Office Supplies      Binders   
3  OFF-ST-10002743  Office Supplies      Storage   
4  FUR-FU-10002116        Furniture  Furnishings   

                                        Product Name    Sales  Quantity  \
0                 Linden 10" Round Wall Clock, Black   48.896         4   
1  Howard Miller 11-1/2" Diameter Brentwood Wall ...  474.430        11   
2            Wilson Jones Easy Flow II Sheet Lifters    3.600         2   
3                      SAFCO Boltless Steel Shelving  454.560         5   
4  Tenex Carpeted, Granite-Look or Clear Contempo...  141.420         5   

   Discount    Profit  
0       0.2    8.5568  
1       0.0  199.2606  
2       0.0    1.7280  
3       0.2 -107.9580  
4       0.6 -187.3815

<small>

### Step 2: Transforming Tabular Data into Text Format

While tabular data is structured and easy for machines to process, it is **not natively understandable by LLMs**. This is because each column in a table exists independently, and LLMs are designed to process continuous text rather than disconnected fields.

To help the LLM "understand" a customer's profile holistically, we convert each row of data into a **single natural-language sentence** that combines all column values.

---

#### 🧾 Example: From Tabular to Text

**Tabular Format (Credit Risk Example)**

| Product Name| Sub-Category | Sales | Discount |
|------------|--------|------------------|----------------|
| Linden 10" Round Wall Clock, Black    | Furnishings | 48.896         | 0.2       |

**Transformed Text Format**

```python
Product Name: Linden 10 Round Wall Clock, Black | Sub-Category: Furnishings | Sales: 48k | Discount: 0,2
```

<small>

In [4]:
# 3. GABUNGKAN KOLOM-KOLOM YANG INFORMATIF UNTUK EMBEDDING
def penggabungan_kolom(df, kolom_data):
    df['teks'] = df[kolom_data].astype(str).agg(' | '.join, axis=1)
    return df

# Gunakan kolom yang relevan
df = penggabungan_kolom(df, ['Product Name', 'Category', 'Sub-Category', 'Sales', 'Discount', 'Profit'])
df

Order Date  Row ID        Order ID       Ship Mode Customer ID  \
0      01-01-20     849  CA-2017-107503  Standard Class    GA-14725   
1      01-01-20    4010  CA-2017-144463  Standard Class    SC-20725   
2      01-01-20    6683  CA-2017-154466     First Class    DP-13390   
3      01-01-20    8070  CA-2017-151750  Standard Class    JM-15250   
4      01-01-20    8071  CA-2017-151750  Standard Class    JM-15250   
...         ...     ...             ...             ...         ...   
3307   30-12-20     908  CA-2017-143259  Standard Class    PO-18865   
3308   30-12-20     909  CA-2017-143259  Standard Class    PO-18865   
3309   30-12-20    1297  CA-2017-115427  Standard Class    EB-13975   
3310   30-12-20    1298  CA-2017-115427  Standard Class    EB-13975   
3311   30-12-20    5092  CA-2017-156720  Standard Class    JM-15580   

          Segment        Country           City       State  Postal Code  \
0        Consumer  United States         Lorain        Ohio        44052   
1        Consumer  United States    Los Angeles  California        90036   
2     Home Office  United States       Franklin   Wisconsin        53132   
3        Consumer  United States     Huntsville       Texas        77340   
4        Consumer  United States     Huntsville       Texas        77340   
...           ...            ...            ...         ...          ...   
3307     Consumer  United States  New York City    New York        10009   
3308     Consumer  United States  New York City    New York        10009   
3309    Corporate  United States      Fairfield  California        94533   
3310    Corporate  United States      Fairfield  California        94533   
3311     Consumer  United States       Loveland    Colorado        80538   

       Region       Product ID         Category Sub-Category  \
0        East  FUR-FU-10003878        Furniture  Furnishings   
1        West  FUR-FU-10001215        Furniture  Furnishings   
2     Central  OFF-BI-10002012  Office Supplies      Binders   
3     Central  OFF-ST-10002743  Office Supplies      Storage   
4     Central  FUR-FU-10002116        Furniture  Furnishings   
...       ...              ...              ...          ...   
3307     East  TEC-PH-10004774       Technology       Phones   
3308     East  OFF-BI-10003684  Office Supplies      Binders   
3309     West  OFF-BI-10002103  Office Supplies      Binders   
3310     West  OFF-BI-10004632  Office Supplies      Binders   
3311     West  OFF-FA-10003472  Office Supplies    Fasteners   

                                           Product Name    Sales  Quantity  \
0                    Linden 10" Round Wall Clock, Black   48.896         4   
1     Howard Miller 11-1/2" Diameter Brentwood Wall ...  474.430        11   
2               Wilson Jones Easy Flow II Sheet Lifters    3.600         2   
3                         SAFCO Boltless Steel Shelving  454.560         5   
4     Tenex Carpeted, Granite-Look or Clear Contempo...  141.420         5   
...                                                 ...      ...       ...   
3307                          Gear Head AU3700S Headset   90.930         7   
3308               Wilson Jones Legal Size Ring Binders   52.776         3   
3309    Cardinal Slant-D Ring Binder, Heavy Gauge Vinyl   13.904         2   
3310                                 GBC Binding covers   20.720         2   
3311                                Bagged Rubber Bands    3.024         3   

      Discount    Profit                                               teks  
0          0.2    8.5568  Linden 10" Round Wall Clock, Black | Furniture...  
1          0.0  199.2606  Howard Miller 11-1/2" Diameter Brentwood Wall ...  
2          0.0    1.7280  Wilson Jones Easy Flow II Sheet Lifters | Offi...  
3          0.2 -107.9580  SAFCO Boltless Steel Shelving | Office Supplie...  
4          0.6 -187.3815  Tenex Carpeted, Granite-Look or Clear Contempo...  
...        ...       ...                                         

<small>

### Step 3: Text Similarity Search

#### Step 3.1: Text Embedding

In this step, the goal is to convert text into a numerical representation (vector) that captures the **meaning or semantics** of the text. This meaning is essential to measure how **similar or close** one data point is to another.

Understanding the similarity between data points is crucial because the primary objective of **Retrieval-Augmented Generation (RAG)** is to **retrieve relevant data** based on a given query, in order to generate **contextually accurate responses** from the LLM.

---

**Model Used: `paraphrase-MiniLM-L6-v2`**

This model is part of the **Sentence Transformers** library, built on top of **BERT**, and is particularly well-suited for **text embedding tasks**. It provides fast and lightweight sentence-level embeddings, making it ideal for semantic search applications like RAG.

Key characteristics:
- Small and fast (~80MB)
- Trained for paraphrase and semantic similarity detection
- Works well with cosine similarity or FAISS indexing

> Tip: replace this with stronger models like `all-mpnet-base-v2` or `e5-large-v2` for bet

<small>

In [5]:
# 4. EMBEDDING DAN NORMALISASI
model = SentenceTransformer('paraphrase-MiniLM-L6-v2')
embedding_dataframe = model.encode(df['teks'].tolist(), convert_to_numpy=True)
embedding_dataframe = embedding_dataframe / np.linalg.norm(embedding_dataframe, axis=1, keepdims=True)
embedding_dataframe = embedding_dataframe.astype('float32')

<small>

#### Step 3.2: FAISS Indexing from Query

**FAISS** (Facebook AI Similarity Search) is a library developed by Facebook AI Research designed for **fast similarity search** over high-dimensional vectors. FAISS is widely used in applications such as **semantic search**, **image similarity**, and **retrieval in RAG (Retrieval-Augmented Generation)**.

---

##### 🛠 How FAISS Works:

1. **Extract vector embeddings** from text using a SentenceTransformer model.
2. **Create and populate the index** with these vectors using FAISS.
3. **Perform nearest neighbor search** based on a query to retrieve the most similar entries.

This indexing and search process enables the system to quickly identify which pieces of information are most relevant to a given query — making it a critical step in any RAG pipeline.

> FAISS supports multiple indexing strategies (e.g., Flat, HNSW, IVFPQ), but `IndexFlatL2` is commonly used for small to medium datasets due to its simplicity and precision.

<small>

In [6]:
# 5. INDEX FAISS
dimension = embedding_dataframe.shape[1]
index = faiss.IndexFlatL2(dimension)
index.add(embedding_dataframe)

In [7]:
# 6. FUNGSI RETRIEVAL
def retrieve(query, index, df, top_k=3):
    query_embedding = model.encode([query], convert_to_numpy=True)
    query_embedding = query_embedding / np.linalg.norm(query_embedding, axis=1, keepdims=True)
    query_embedding = query_embedding.astype("float32")
    scores, indices = index.search(query_embedding, top_k)
    result_df = df.iloc[indices[0]].copy()
    result_df['similarity_score'] = scores[0]
    return result_df

In [8]:
# 7. FUNGSI GENERATE JAWABAN
def generate_answer(query, context, api_key):
    openai.api_key = api_key
    system_message = "Kamu adalah asisten cerdas yang menjawab pertanyaan berdasarkan data yang diberikan."
    user_message = f"""
    Pertanyaan: {query}
    Data yang relevan:
    {context}
    """
    response = openai.ChatCompletion.create(
        model="gpt-4-1106-preview",  # gunakan model yang tersedia
        messages=[
            {"role": "system", "content": system_message},
            {"role": "user", "content": user_message}
        ],
        temperature=0.3,
        max_tokens=1000
    )
    return response.choices[0].message["content"]

In [ ]:
# 8. CONTOH PENGGUNAAN
query = "Produk apa yang memiliki diskon tertinggi?"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Berdasarkan data yang diberikan, diskon untuk setiap produk dihitung sebagai persentase dari harga sebelum diskon. Produk dengan diskon tertinggi adalah produk dengan persentase diskon terbesar.

Berikut adalah perhitungan diskon untuk setiap produk:

1. Okidata B400 Printer:
   Harga sebelum diskon: $343.2
   Diskon: 20% (0.2)
   Jumlah diskon: $343.2 * 0.2 = $68.64

2. Okidata MB491 Multifunction Printer:
   Harga sebelum diskon: $2395.2
   Diskon: 20% (0.2)
   Jumlah diskon: $2395.2 * 0.2 = $479.04

3. Okidata C610n Printer:
   Harga sebelum diskon: $649.0
   Diskon: 50% (0.5)
   Jumlah diskon: $649.0 * 0.5 = $324.5

Dari perhitungan tersebut, produk dengan diskon tertinggi adalah Okidata C610n Printer dengan diskon 50% atau $324.5.


In [12]:
query = "Produk mana yang memberikan profit negatif?"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Untuk menentukan produk mana yang memberikan profit negatif, kita perlu melihat pada data yang diberikan dan menghitung profit untuk setiap produk. Profit dihitung dengan mengurangkan biaya dari harga jual. Jika hasilnya negatif, maka produk tersebut memberikan profit negatif.

Dari data yang diberikan:

1. Zebra Zazzle Fluorescent Highlighters (pertama)
   Harga Jual: $9.728
   Diskon: 0.2 (20% dari harga jual)
   Biaya: $1.7024

   Profit = (Harga Jual - Diskon) - Biaya
   Profit = ($9.728 - $9.728 * 0.2) - $1.7024
   Profit = ($9.728 - $1.9456) - $1.7024
   Profit = $7.7824 - $1.7024
   Profit = $6.08

2. Eureka The Boss Cordless Rechargeable Stick Vac
   Harga Jual: $152.94
   Diskon: 0.0 (tidak ada diskon)
   Biaya: $41.2938

   Profit = (Harga Jual - Diskon) - Biaya
   Profit = ($152.94 - $0) - $41.2938
   Profit = $152.94 - $41.2938
   Profit = $111.6462

3. Zebra Zazzle Fluorescent Highlighters (kedua)
   Harga Jual: $14.592
   Diskon: 0.2 (20% dari harga jual)
   Biaya: $2.553

In [13]:
query = "Kategori apa yang paling sering muncul?"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Berdasarkan data yang relevan yang Anda berikan, terdapat tiga kategori yang disebutkan:

1. Furniture
2. Office Supplies
3. Office Supplies

Dari ketiga kategori tersebut, "Office Supplies" muncul dua kali, sedangkan "Furniture" muncul satu kali. Oleh karena itu, kategori yang paling sering muncul dalam data yang diberikan adalah "Office Supplies".


In [14]:
query = "Produk dengan harga tertinggi?"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Berdasarkan data yang diberikan, produk "Eureka The Boss Cordless Rechargeable Stick Vac" memiliki harga tertinggi sebesar $254.90.


In [15]:
query = "Produk apa yang paling sering rugi?"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Untuk menentukan produk yang paling sering rugi, kita perlu melihat margin keuntungan dari setiap produk. Margin keuntungan dapat dihitung dengan mengurangkan biaya dari harga jual, dan kemudian membandingkan dengan harga jual untuk mendapatkan persentase margin keuntungan.

Berdasarkan data yang diberikan:

1. Okidata B400 Printer
   Harga Jual: $343.2
   Diskon: 0.2 (20% dari harga jual)
   Biaya: $38.61

   Harga Jual setelah Diskon = Harga Jual - (Harga Jual * Diskon)
   Harga Jual setelah Diskon = $343.2 - ($343.2 * 0.2)
   Harga Jual setelah Diskon = $343.2 - $68.64
   Harga Jual setelah Diskon = $274.56

   Margin Keuntungan = Harga Jual setelah Diskon - Biaya
   Margin Keuntungan = $274.56 - $38.61
   Margin Keuntungan = $235.95

2. Carina Mini System Audio Rack, Model AR050B
   Harga Jual: $332.94
   Diskon: 0.0 (Tidak ada diskon)
   Biaya: $9.9882

   Harga Jual setelah Diskon = Harga Jual - (Harga Jual * Diskon)
   Harga Jual setelah Diskon = $332.94 - ($332.94 * 0.0)
   Har

In [16]:
query = "produk dengan diskon tinggi tapi tetap untung"
top_result = retrieve(query, index, df, top_k=3)
context = "\n".join(top_result['teks'].tolist())
# API key kamu masukkan di parameter
print(generate_answer(query, context, api_key=("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")))

Berdasarkan data yang diberikan, kita memiliki dua produk yang sama ("Eureka The Boss Cordless Rechargeable Stick Vac") dengan harga dan margin keuntungan yang berbeda, dan satu produk lainnya ("Konftel 250 Conference phone - Charcoal black") dengan diskon.

Untuk menentukan produk mana yang memiliki diskon tinggi tetapi tetap menguntungkan, kita perlu melihat diskon yang diberikan dan margin keuntungan dari setiap produk.

1. Eureka The Boss Cordless Rechargeable Stick Vac (pertama):
   - Harga: $152.94
   - Diskon: 0.0% (tidak ada diskon)
   - Margin Keuntungan: $41.2938

2. Konftel 250 Conference phone - Charcoal black:
   - Harga: $455.712
   - Diskon: 20.0%
   - Margin Keuntungan: $34.1784

3. Eureka The Boss Cordless Rechargeable Stick Vac (kedua):
   - Harga: $254.9
   - Diskon: 0.0% (tidak ada diskon)
   - Margin Keuntungan: $68.823

Dari data di atas, hanya "Konftel 250 Conference phone - Charcoal black" yang memiliki diskon (20.0%). Meskipun diskon ini tinggi, produk ini masi